internal/peer/pubsub.go

```diff
@@ -0,0 +1,269 @@
+package peer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/dgryski/go-wyhash"
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/generics"
+	"github.com/honeycombio/refinery/pubsub"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// refreshCacheInterval is how frequently this host will re-register itself
+	// with Redis. This should happen about 3x during each timeout phase in order
+	// to allow multiple timeouts to fail and yet still keep the host in the mix.
+	// Falling out of Redis will result in re-hashing the host-trace affinity and
+	// will cause broken traces for those that fall on both sides of the rehashing.
+	// This is why it's important to ensure hosts stay in the pool.
+	refreshCacheInterval = 3 * time.Second
+
+	// peerEntryTimeout is how long redis will wait before expiring a peer that
+	// doesn't check in. The ratio of refresh to peer timeout should be 1/3. Redis
+	// timeouts are in seconds and entries can last up to 2 seconds longer than
+	// their expected timeout (in my load testing), so the lower bound for this
+	// timer should be ... 5sec?
+	peerEntryTimeout = 10 * time.Second
+)
+
+type peerAction string
+
+const (
+	Register   peerAction = "R"
+	Unregister peerAction = "U"
+)
+
+type peerCommand struct {
+	action peerAction
+	peer   string
+}
+
+func newPeerCommand(action peerAction, peer string) *peerCommand {
+	return &peerCommand{
+		action: action,
+		peer:   peer,
+	}
+}
+
+func (p *peerCommand) unmarshal(msg string) bool {
+	if len(msg) < 2 {
+		return false
+	}
+	p.action = peerAction(msg[:1])
+	p.peer = msg[1:]
+	switch p.action {
+	case Register, Unregister:
+		return true
+	default:
+		return false
+	}
+}
+
+func (p *peerCommand) marshal() string {
+	return string(p.action) + p.peer
+}
+
+type pubsubPeers struct {
+	Config config.Config   `inject:""`
+	PubSub pubsub.PubSub   `inject:""`
+	Clock  clockwork.Clock `inject:""`
+
+	peers     *generics.SetWithTTL[string]
+	hash      uint64
+	callbacks []func()
+	sub       pubsub.Subscription
+	done      chan struct{}
+}
+
+// NewPubsubPeers returns a peers collection backed by a pubsub system.
+// It expects members to publish their own presence to the pubsub system every so often
+// as determined by refreshCacheInterval. If they fail to do after peerEntryTimeout, they
+// will be removed from the list of peers.
+// They can also remove themselves from the list of peers by publishing an "unregister" message.
+// The register message is just "register address" and the unregister message is "unregister address".
+func newPubsubPeers(c config.Config) (Peers, error) {
+	return &pubsubPeers{
+		Config: c,
+		Clock:  clockwork.NewRealClock(),
```

Comment:
I don't think we want to instantiate a new clock instance if `Clock` is injected